### PR TITLE
perf(review): bound the stale-literal signal scan to a single repo pass

### PR DESCRIPTION
**Pull request by alfhen**
## What

`computeStaleLiteralCandidates` (via the old `findStaleSites`) re-walked every
line of every indexed repo chunk once **per distinct touched literal**. The
rule triggers unconditionally (`always: true`) and is rendered before the LLM
even runs, so this cost — O(literals × repo-lines) — was paid on every single
PR review, including the common case where none of the touched literals
survive elsewhere in the repo (the only early exit, `MAX_SITES_PER_LITERAL`,
fires after matches are found).

## Why

Adversarially-confirmed review finding: no time budget bounded this scan,
unlike `grep_codebase` in the same package (`GREP_TIME_BUDGET_MS`), and
`chunk.content.split('\n')` was being redone once per literal per chunk
instead of once per chunk.

## Approach

- Restructured to a single repo-wide pass: `scanRepoForStaleSites` builds
  `literal value -> surviving sites` directly while walking `repoChunks`
  exactly once, using a per-line quoted-token extraction (`extractQuotedValues`)
  checked against the small set of touched literal values — O(repo-lines +
  literals) instead of O(literals × repo-lines).
- `MAX_SITES_PER_LITERAL`'s "first-N in traversal order, per literal" semantics
  are preserved exactly: sites are still capped independently per literal
  value, in chunk-then-line order.
- Added `STALE_LITERAL_TIME_BUDGET_MS` (5s, mirroring `GREP_TIME_BUDGET_MS`)
  as a wall-clock backstop. On budget hit, the scan returns whatever it found
  so far and logs a diagnostic via `context.logger.warning(...)` (chunks
  scanned, elapsed ms, touched-literal count) so a slow review is
  diagnosable instead of silently truncated.
- No prompt or rule changes. The rendered `<stale_literal_candidates>` /
  "None" block format is untouched, and output is identical to before for any
  input that completes within budget — verified by a multi-literal,
  multi-file regression fixture (see tests below). One accepted, documented
  edge case: because the new path extracts proper quote-delimited tokens
  rather than doing a raw substring search, a literal value nested inside a
  *differently*-quoted outer string (e.g. inside a backtick template) is no
  longer matched as a separate token. This wasn't exercised by any existing
  test/fixture and is called out in a code comment.
- `computeStaleLiteralCandidates(context)` keeps its existing signature/
  behavior; a new `computeStaleLiteralCandidatesWithDeadline(context, deadline)`
  is exported alongside it purely so the budget-exceeded path can be tested
  deterministically (inject an already-past deadline) without any reliance on
  real elapsed time.

## Verification

```
npm run format:check   # pass
npm run lint           # pass (0 errors; pre-existing warnings only)
npm run typecheck      # @liendev/review typecheck: pass (clean)
npm run build          # pass (parser, core, review, cli, action)
npm run test -w @liendev/review   # 458/458 passed (25 files), incl. 27/27 in
                                    # stale-literal-signals.test.ts
```

Added tests (all new, existing 23 unmodified and green):
- Multi-literal/multi-file regression fixture proving output equivalence,
  including the per-literal `MAX_SITES_PER_LITERAL` cap and first-N-in-
  traversal-order selection.
- A counting-chunk structural test proving each repo chunk's `.content` is
  read exactly once regardless of how many literals are touched (would have
  been `chunks × literals` reads under the old per-literal scan).
- Budget-hit path: an injected already-past deadline deterministically
  exercises the timeout branch, asserting partial (empty) results plus
  exactly one logged warning diagnostic — no timing-based flakiness.
- A budget-not-hit control test asserting no diagnostic is logged on the
  normal path.

## Caveats

- `npm run typecheck` (root, chains all workspaces) currently fails on
  `packages/cli` with `Property 'removeFiles' does not exist on type
  'ManifestManager'`. This is **pre-existing and unrelated** to this change:
  this worktree's `node_modules` (per the documented worktree symlink
  workaround) resolves `@liendev/core` back to the main checkout's
  `packages/core/dist`, which is stale relative to main's own already-merged
  source (the manifest-batching PR added `removeFiles`, but that checkout's
  `dist/` hasn't been rebuilt). `git diff` confirms zero changes to
  `packages/cli` in this PR, and `@liendev/review`'s own `tsc --noEmit` passes
  cleanly. `npm run build` (root, tsup-based) also passes end to end.
- `@liendev/review` is private (no changeset needed).

---

### Lien Review

✅ **Improved!** Complexity reduced by 15. 1 pre-existing issue remains in touched files.
| Metric | Violations | Change |
|--------|:----------:|:------:|
| 🧠 mental load | 1 | -15 |

> [!NOTE]
> **Low Risk**
>
> This PR restructures the stale-literal survivor scan into a single repo-wide pass with a 5s wall-clock budget, preserving the existing per-literal cap and traversal-order semantics. It adds explicit 'incomplete scan' rendering so the agent is not told discovery is complete when the budget aborts the scan, and reintroduces line/value deduplication across overlapping chunks. No exports were removed or changed in signature, and the stricter whole-token matching is intentionally scoped and documented.
>
> - Single-pass survivor scan via scanRepoForStaleSites and extractQuotedValues, replacing O(literals × repo-lines) per-literal walk
> - Added STALE_LITERAL_TIME_BUDGET_MS and deterministic deadline entry points for timeout testing
> - Distinguishes a clean zero-match scan from a budget-aborted scan in renderStaleLiteralSection

<sup>Reviewed by [Lien Review](https://lien.dev) for commit 03cee34. Updates automatically on new commits.</sup>

### CODE REVIEW

**Comment by alfhen** [comment]
## Complexity follow-up (commit 2e8a92b)

Lien's own review flagged an average-complexity increase from the single-pass restructure. Measured directly with `@liendev/parser`'s `analyzeComplexityFromChunks` / per-chunk cyclomatic+cognitive+Halstead metadata, before and after:

| Function | Metric | Before (this PR's HEAD) | After refactor | Threshold |
|---|---|---:|---:|---:|
| `scanRepoForStaleSites` | cognitive | 29 | 9 | 15 |
| `scanRepoForStaleSites` | halstead (time to understand) | ~1h 11m | ~30m | 1h |
| `computeStaleLiteralCandidatesWithDeadline` | cognitive | 11 | 6 | 15 |
| `computeStaleLiteralCandidatesWithDeadline` | halstead (time to understand) | ~1h 11m | ~23m | 1h |

File-level (cyclomatic):

| | avgComplexity | maxComplexity | violations |
|---|---:|---:|---:|
| main (pre-PR) | 5.0 | 9 | 2 |
| PR HEAD (e8f7bdb) | 5.3 | 11 | 4 |
| after refactor | 4.5 | 9 | 1 |

The 1 remaining violation (`scanPatches`, cognitive 22) is pre-existing and untouched by this PR — out of scope here.

**What changed:** extracted `recordSurvivorsForLine` (per-line quoted-value matching + per-literal cap, out of the triple-nested scan loop), `buildRankedCandidates` (candidate scoring/sort/cap), and `logScanTimeout` (diagnostic message) out of the two flagged functions. Pure extraction — no exported signatures, emitted signal shape/ordering/caps, or test expectations changed.

Verification: `format:check` / `lint` / `typecheck` / `build` all pass; `npm run test -w @liendev/review` — 458/458 passed (25 files), including all 27/27 in `stale-literal-signals.test.ts` unmodified.

File: packages/review/src/stale-literal-signals.ts [hunk 1]
```diff
@@ -241,45 +242,108 @@ function classifySnippet(line: string): { snippet: string; isComment: boolean }
...
+ *
+ * This replaces a previous per-literal scan (call this once per touched
+ * literal, re-walking every repo line each time), which made the cost
+ * O(literals x repo-lines) — unconditional on every PR, with the common
+ * zero-match case paying full cost. A single pass makes it
+ * O(repo-lines + literals).
  */
-function findStaleSites(
-  lit: ChangedLiteral,
+function scanRepoForStaleSites(
```
[the remark this answers is not in the document]
Comment (alfhen) [reply packages/review/src/stale-literal-signals.ts:328]: Fixed in 03cee34: reintroduced a `seenSites` dedup set keyed by `${file}:${absLine}:${value}` in `recordSurvivorsForLine`/`scanRepoForStaleSites`. Already-recorded sites are now skipped even when repoChunks contain overlapping ranges for the same file, so a survivor line is never counted twice against `MAX_SITES_PER_LITERAL` or confidence scoring.

Added a regression test (`dedupes survivors when repoChunks contain overlapping ranges for the same file`) with two chunks that both cover the same physical line, asserting exactly one recorded site.

**Comment by alfhen** [comment]
Also fixed the second finding from the review summary (timeout lies), which was outside the diff so it had no inline thread: `renderStaleLiteralSection` discarded the `timedOut` signal from the scan, so a scan that hit its time budget before finding any survivors still rendered the "None — ... discovery step is complete" block, telling the agent it need not grep exactly when the scan was incomplete.

Fix (03cee34): the scan and render paths now share an internal `computeStaleLiteralScan` that surfaces `timedOut` (plus chunk coverage) alongside the candidate list. `renderStaleLiteralSection` renders an explicit `Scan incomplete (time budget exceeded after N of M chunks) — ... grep for the diff's changed literals yourself` notice instead of the "None" claim when the budget was exceeded before any survivor was found.

Scope note: only the timeout-path rendering changed. Output for scans that complete within budget is byte-identical to before — verified by the existing 27 tests (unmodified) plus a new "does not log / returns candidates unchanged" style check. Added:
- a timeout-with-zero-survivors test asserting the incomplete-scan notice replaces the "None" block
- a timeout-with-partial-survivors test proving the normal candidate block still renders unchanged in that case (deliberate: the candidate block never claims completeness, so only the zero-candidate "None" claim was ever misleading)

All gates green: `format:check`, `lint` (0 errors), `typecheck`, `build`, and `npm run test -w @liendev/review` (461/461, incl. 30/30 in `stale-literal-signals.test.ts`, up from 27).

Pushed as 03cee34. If you want belt-and-braces on top of the unit tests, a CC-mode harness replay (`/test-harness stale-duplicate` or similar) would be worth a quick sanity check since this module's output is prompt-visible, though no calibration-sensitive text changed on the non-timeout path.
